### PR TITLE
Add camera video recorder system

### PR DESCRIPTION
**Pull request by iche033**
Add camera video recorder system that can be attached to a camera sensor and provides a service for recording a video from the sensor's image stream.

To test, launch the `camera_video_record_dbl_pendulum.sdf` world:

```
ign gazebo -v 4 -r ./src/ign-gazebo/examples/worlds/camera_video_record_dbl_pendulum.sdf
```
Call the following service to start video recording:

```
ign service -s /camera/record_video --reqtype ignition.msgs.VideoRecord --reptype ignition.msgs.Boolean --timeout 300 --req 'start: true, format:"mp4", save_filename:"test.mp4"'
```

Call the following service to stop video recording:

```
ign service -s /camera/record_video --reqtype ignition.msgs.VideoRecord --reptype ignition.msgs.Boolean --timeout 300 --req 'stop: true'
```

The video will be saved to `test.mp4`

### CODE REVIEW

File: examples/worlds/camera_video_record_dbl_pendulum.sdf [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+      filename="libignition-gazebo-scene-broadcaster-system.so"
+      name="ignition::gazebo::systems::SceneBroadcaster">
+    </plugin>
+    <plugin
+      filename="libignition-gazebo-sensors-system.so"
+      name="ignition::gazebo::systems::Sensors">
+      <render_engine>ogre2</render_engine>
+    </plugin>
+
+    <gui fullscreen="0">
```
Comment (nkoenig) [inline examples/worlds/camera_video_record_dbl_pendulum.sdf:42]: I think you can remove the `<gui>` tags now that #317 is in.

File: src/systems/camera_video_recorder/CameraVideoRecorder.cc [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+    this->recordVideoFormat = _msg.format();
+    this->recordVideoSavePath = _msg.save_filename();
+
+    if (this->recordVideoFormat.empty())
+      this->recordVideoFormat = "mp4";
+
+    // create filename with timestamped suffix if path is not specified
+    if (this->recordVideoSavePath.empty())
+    {
+      std::chrono::system_clock::time_point p =
```
Comment (nkoenig) [inline src/systems/camera_video_recorder/CameraVideoRecorder.cc:137]: Can you use: `std::string str = common::systemTimeISO();`

File: src/systems/camera_video_recorder/CameraVideoRecorder.cc [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+//////////////////////////////////////////////////
+void CameraVideoRecorder::Configure(
+    const Entity &_entity, const std::shared_ptr<const sdf::Element> &_sdf,
+    EntityComponentManager &_ecm, EventManager &_eventMgr)
+{
+  auto cameraEntComp =
+      _ecm.Component<components::Camera>(_entity);
+  if (!cameraEntComp)
+  {
+    ignerr << "The Camera video recorder system can only be attached to a "
```
Comment (nkoenig) [inline src/systems/camera_video_recorder/CameraVideoRecorder.cc:184]:
```suggestion
    ignerr << "The camera video recorder system can only be attached to a "
```

File: src/systems/camera_video_recorder/CameraVideoRecorder.cc [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+           << "camera sensor" << std::endl;
```
Comment (nkoenig) [inline src/systems/camera_video_recorder/CameraVideoRecorder.cc:185]:
```suggestion
           << "camera sensor." << std::endl;
```

File: src/systems/camera_video_recorder/CameraVideoRecorder.hh [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+namespace systems
+{
+  class CameraVideoRecorderPrivate;
+
+  /** \class CameraVideoRecorder CameraVideoRecorder.hh \
+   * ignition/gazebo/systems/CameraVideoRecorder.hh
+  **/
+  /// \brief Record video from a camera sensor
+  /// The system takes in the following parameter:
+  ///   <topic>    Name of topic for the video recorder service. If this is
```
Comment (nkoenig) [inline src/systems/camera_video_recorder/CameraVideoRecorder.hh:40]: Should we call this `<service>` or `<service_topic>`, instead of `<topic>`?

**Comment by nkoenig** [comment]
Works great. Just a few comments.

File: src/systems/camera_video_recorder/CameraVideoRecorder.hh (hunk 5 above)
Comment (iche033) [reply src/systems/camera_video_recorder/CameraVideoRecorder.hh:40]: renamed to `<service>` d90f78e

File: src/systems/camera_video_recorder/CameraVideoRecorder.cc (hunk 4 above)
Comment (iche033) [reply src/systems/camera_video_recorder/CameraVideoRecorder.cc]: fixed. d90f78e

File: src/systems/camera_video_recorder/CameraVideoRecorder.cc (hunk 3 above)
Comment (iche033) [reply src/systems/camera_video_recorder/CameraVideoRecorder.cc]: fixed. d90f78e

File: src/systems/camera_video_recorder/CameraVideoRecorder.cc (hunk 2 above)
Comment (iche033) [reply src/systems/camera_video_recorder/CameraVideoRecorder.cc]: done. d90f78e

File: examples/worlds/camera_video_record_dbl_pendulum.sdf (hunk 1 above)
Comment (iche033) [reply examples/worlds/camera_video_record_dbl_pendulum.sdf]: removed <gui>. d90f78e

**Comment by iche033** [comment]
> Only nit would be: Is there a way to expose available video formats, or at least give a warning when there is an incorrect format?

added check and print out error if format not supported, and also lists available formats. ed72b72

**Review by chapulina — CHANGES_REQUESTED** [review]
The new test is failing on Linux and on Mac:

```
/var/lib/jenkins/workspace/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/ign-gazebo/test/integration/camera_video_record_system.cc:80
Value of: hasService
  Actual: false
Expected: true
```

```
53: [ RUN      ] CameraVideoRecorderTest.RecordVideo
53: [Wrn] [URI.cc:580] Unable to parse URI [/Users/jenkins/workspace/ignition_gazebo-ci-pr_any-homebrew-amd64/ign-gazebo/test/worlds/camera_video_record.sdf]. Ignoring.
53: objc[52149]: Class OgreConfigWindowDelegate is implemented in both /usr/local/opt/ogre1.9/lib/libOgreMain.1.9.0.dylib (0x116d86dc0) and /usr/local/opt/ogre2.1/lib/libOgreMain.2.1.0.dylib (0x1167af838). One of the two will be used. Which one is undefined.
53: dyld: lazy symbol binding failed: Symbol not found: __ZN4Ogre2v113OverlaySystemC1Ev
53:   Referenced from: /usr/local/Cellar/ignition-rendering3/3.1.0/lib/libignition-rendering3-ogre2.dylib
53:   Expected in: flat namespace
53: 
53: dyld: Symbol not found: __ZN4Ogre2v113OverlaySystemC1Ev
53:   Referenced from: /usr/local/Cellar/ignition-rendering3/3.1.0/lib/libignition-rendering3-ogre2.dylib
53:   Expected in: flat namespace
53: 
 53/138 Test  #53: INTEGRATION_camera_video_record_system ...............Child aborted***Exception:   1.18 sec
```

**Comment by iche033** [comment]
@osrf-jenkins run tests please

**Comment by iche033** [comment]
@osrf-jenkins run tests please

**Comment by iche033** [comment]
latest CI builds look good. Failing tests on homebrew are unrelated.

**Review by chapulina — COMMENTED** [review]
The test failures have been fixed, so I can remove my request for changes. I haven't reviewed the PR, but the 2 approvals are enough.

**Review by chapulina — APPROVED** [review]

**Pull request merged**